source/scripts/model/DungeonGenerator.js

```diff
@@ -0,0 +1,504 @@
+ * Encapsulates the space that is ultimately part of the dungeon.
+ * @todo Ensure that the intersects method actually works...
+ */
+class Agent {
+    /**
+     * Creates an agent.
+     * @param {Space} space - Space that this agent encapsulates.
+     */
+    constructor(space) {
+        this.space = space
+        this.velocity = {x: 0, y: 0}
+        this.delete = false
+    }
+    /**
+     * Determines the central point of the encapsulated space.
+     * @return {x, y} Object containing x & y coordinates.
+     */
+    centroid() {
+        return {
+            x: this.space.position.x + this.space.width / 2,
+            y: this.space.position.y + this.space.height / 2
+        }
+    }
+    /**
+     * Calculates the square of the distance between two agents.
+     * @param {Agent} other - Agent from which to determine distance.
+     * @return {number} square distance
+     */
+    distanceSquare(other) {
+        let center1 = this.centroid()
+        let center2 = other.centroid()
+
+        return Math.pow(center1.x - center2.x, 2) +
+            Math.pow(center1.y - center2.y, 2)
+    }
+    /**
+     * Determines if another agent intersects this one.
+     * @param {Agent} agent - Agent test
+     * @return {bool} true if the other agent intersects this agent.
+     */
+    intersects(agent) {
+        let r1 = {
+            left: this.space.position.x,
+            top: this.space.position.y,
+            right: this.space.position.x + this.space.width,
+            bottom: this.space.position.y + this.space.height
+        }
+
+        let r2 = {
+            left: agent.space.position.x,
+            top: agent.space.position.y,
+            right: agent.space.position.x + agent.space.width,
+            bottom: agent.space.position.y + agent.space.height
+        }
+
+        return (r1.left < r2.right &&
+            r1.right > r2.left &&
+            r1.top > r2.bottom &&
+            r1.bottom < r2.top )
+
+        //return !(r2.left > r1.right ||
+        //   r2.right < r1.left ||
+        //   r2.top > r1.bottom ||
+        //   r2.bottom < r1.top);
+	}
+}
+
+/**
+ * Class to generate a random dungeon.
+ * @todo Fix trimming
+ * @todo Make more robust against islands
+ */
+class DungeonGenerator {
+    /**
+     * Constructs a random dungeon generator.
+     * @param {integer} maxSpaces - Maximum number of random spaces to generate.
+     * @param {width, height} minSpace - Minimum dimensions for a space.
+     * @param {width, height} maxSpace - Maximum dimensions for a space.
+     */
+    constructor(maxSpaces = null, minSpace = null, maxSpace = null) {
+        // arbitrary default value because I'm testing
+        this.maximumSpaces = maxSpaces || 30
+        this.minimumSpace = minSpace || {
+            width: 3,
+            height: 3
+        }
+        this.maximumSpace = maxSpace || {
+            width: 20,
+            height: 20
+        }
```

Comment:
no reason for the first one, just tired--I put it in there as a default

the rest, yeah, they long and it would've been ugly lol